src/rgw/driver/posix/rgw_sal_posix.cc

```diff
@@ -0,0 +1,3090 @@
+  std::string dst_fname = to->get_fname();
+  int flags = 0;
+
+  if (to->exists(dpp)) {
+    flags = RENAME_EXCHANGE;
+  }
+  // swap
+  int ret = renameat2(tb->get_dir_fd(dpp), src_fname.c_str(), tb->get_dir_fd(dpp), dst_fname.c_str(), flags);
+  if(ret < 0) {
+    ret = errno;
+    ldpp_dout(dpp, 0) << "ERROR: renameat2 for shadow object could not finish: "
+	<< cpp_strerror(ret) << dendl;
+    return -ret;
+  }
+
+  // Update saved bucket info
+  info.bucket.name = to->get_name();
+  bufferlist bl;
+  encode(info, bl);
+  ret = write_x_attr(dpp, y, dir_fd, RGW_POSIX_ATTR_BUCKET_INFO, bl, get_name());
+  if (ret < 0) {
+    return ret;
+  }
+
+  // Delete old one (could be file or directory)
+  struct statx stx;
+  ret = statx(parent_fd, src_fname.c_str(), AT_SYMLINK_NOFOLLOW,
+		  STATX_ALL, &stx);
+  if (ret < 0) {
+    ret = errno;
+    if (ret == ENOENT) {
+      return 0;
+    }
+    ldpp_dout(dpp, 0) << "ERROR: could not stat object " << get_name() << ": "
+                  << cpp_strerror(ret) << dendl;
+    return -ret;
+  }
+
+  if (S_ISREG(stx.stx_mode)) {
+    ret = unlinkat(parent_fd, src_fname.c_str(), 0);
+  } else if (S_ISDIR(stx.stx_mode)) {
+    ret = delete_directory(parent_fd, src_fname.c_str(), true, dpp);
+  }
+  if (ret < 0) {
+    ret = errno;
+    ldpp_dout(dpp, 0) << "ERROR: could not remove old file " << get_name()
+                      << ": " << cpp_strerror(ret) << dendl;
+    return -ret;
+  }
+
+  return 0;
+}
+
+int POSIXBucket::close()
+{
+  if (dir_fd < 0) {
+    return 0;
+  }
+
+  ::close(dir_fd);
+  dir_fd = -1;
+
+  return 0;
+}
+
+int POSIXBucket::stat(const DoutPrefixProvider* dpp)
+{
+  if (stat_done) {
+    return 0;
+  }
+
+  int ret = statx(parent_fd, get_fname().c_str(), AT_SYMLINK_NOFOLLOW,
+		  STATX_ALL, &stx);
+  if (ret < 0) {
+    ret = errno;
+    ldpp_dout(dpp, 0) << "ERROR: could not stat bucket " << get_name() << ": "
+                  << cpp_strerror(ret) << dendl;
+    return -ret;
+  }
+  if (!S_ISDIR(stx.stx_mode)) {
+    /* Not a bucket */
+    return -EINVAL;
+  }
+
+  stat_done = true;
+  return 0;
+}
+
+/* This is a shadow bucket.  Copy it into a new shadow bucket in the destination
+ * bucket */
+int POSIXBucket::copy(const DoutPrefixProvider *dpp, optional_yield y,
+                      POSIXBucket* db, POSIXObject* dest)
+{
+  std::unique_ptr<POSIXBucket> dsb;
+
+  // Delete the target, in case it's not a multipart
+  int ret = dest->delete_object(dpp, y);
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: could not remove dest object "
+                      << dest->get_name() << dendl;
+    return ret;
+  }
+
+  ret = db->get_shadow_bucket(dpp, y, std::string(), std::string(), dest->get_fname(), true, &dsb);
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: could not create shadow bucket " << dest->get_name()
+                      << " in bucket " << db->get_name() << dendl;
+    return ret;
+  }
+
+  for_each(dpp, [this, &dsb, &dpp, &y](const char *name) {
```

Comment:
how are intermediate errors communicated back to the caller?

Reply:
will fix.